src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+}
+
+// This function is called by the interaction model engine when a command destined for this instance is received.
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case ModeBase::Commands::ChangeToMode::Id:
+        ChipLogDetail(Zcl, "ModeBase: Entering handling ChangeToModeWithStatus");
+
+        HandleCommand<Commands::ChangeToMode::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { handleChangeToMode(ctx, commandData); });
+    }
+}
+
+// List the commands supported by this instance.
+CHIP_ERROR Instance::EnumerateAcceptedCommands(const ConcreteClusterPath & cluster,
+                                               CommandHandlerInterface::CommandIdCallback callback, void * context)
+{
+    callback(ModeBase::Commands::ChangeToMode::Id, context);
+    callback(ModeBase::Commands::ChangeToModeResponse::Id, context);
+    return CHIP_NO_ERROR;
+}
+
+// Implements the read functionality for complex attributes.
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Attributes::CurrentMode::Id:
+        ReturnErrorOnFailure(aEncoder.Encode(mCurrentMode));
+        break;
+    case Attributes::StartUpMode::Id:
+        if (mStartUpMode.IsNull())
+        {
+            ReturnErrorOnFailure(aEncoder.EncodeNull());
+        }
+        else
+        {
+            ReturnErrorOnFailure(aEncoder.Encode(mStartUpMode));
+        }
+        break;
+    case Attributes::OnMode::Id:
+        if (mOnMode.IsNull())
+        {
+            ReturnErrorOnFailure(aEncoder.EncodeNull());
+        }
+        else
+        {
+            ReturnErrorOnFailure(aEncoder.Encode(mOnMode));
+        }
+        break;
+    case Attributes::SupportedModes::Id:
+        if (NumberOfModes() == 0)
+        {
+            aEncoder.EncodeEmptyList();
+            return CHIP_NO_ERROR;
+        }
+
+        Instance * d   = this;
+        CHIP_ERROR err = aEncoder.EncodeList([d](const auto & encoder) -> CHIP_ERROR {
+            for (uint8_t i = 0; i < d->NumberOfModes(); i++)
+            {
+                ModeOptionStructType mode;
+                ChipError err1 = CHIP_NO_ERROR;
+
+                // Get the mode label
+                char buffer[64];
+                MutableCharSpan label(buffer);
+                err1 = d->GetModeLabelByIndex(i, label);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+                mode.label = label;
+
+                // Get the mode value
+                err1 = d->GetModeValueByIndex(i, mode.mode);
+                if (err1 != CHIP_NO_ERROR)
+                {
+                    return err1;
+                }
+
+                // Get the mode tags
+                ModeTagStructType tagsBuffer[8];
```

Comment:
GetModeTagsByIndex needs to take a bound so we don't stack-overrun here!

Reply:
The caller here is the SDK. It is explicitly mentioned in the spec and in the documentation that the buffer provided **will** be "of sufficient size to hold the spec-required number of tags". This gives the callee all the space required.